components/avatar/avatar.vue

```diff
@@ -125,33 +147,98 @@ export default {
   },
 
   methods: {
-    init () {
-      const firstChild = this.$el.firstChild;
-      if (firstChild) {
-        this.setKind(firstChild);
-        if (this.kind === 'image') {
-          firstChild.classList.add('d-avatar__image');
+    async init () {
+      const firstChild = this.$slots.default[0]?.elm || this.$slots.default[0];
+      const iconChild = this.$slots.default[0]?.componentOptions?.tag === 'dt-icon';
+
+      if (firstChild || iconChild) {
+        if (iconChild) {
+          this.kind = 'icon';
+        } else {
+          this.setKind(firstChild);
         }
+
+        await this.$nextTick();
+        const childEl = this.$el;
+
+        if (this.kind === 'image' && this.imageLoadedSuccessfully === null) {
+          childEl.firstChild.classList.add('d-avatar__image');
+
+          childEl.firstChild.addEventListener('error', () => {
+            this.kind = 'initials';
+            this.imageLoadedSuccessfully = false;
+            childEl.firstChild?.classList?.remove('d-avatar__image');
+          });
+
+          childEl.firstChild.addEventListener('load', () => {
+            this.imageLoadedSuccessfully = true;
+            childEl.classList.add('d-avatar__image--loaded');
+          });
```

Comment:
I initially was thinking you would change the `--avatar-color-background` and the `--avatar-gradient-stop-N` to `transparent`, but I like this better.

Nitpick: BEM structure is a bit off for `d-avatar__image--loaded`, since `--loaded` is not directly modifying the `img.d-avatar__image`. 

I'd go with `d-avatar--image-loaded`.

Reply:
Yes it's better with `d-avatar--image-loaded`, updated